File: algorithms/leetcode/791-custom-sort-string/CustomSortString.java
class Solution {
    public String customSortString(String S, String T) {
        int[] pos = new int[26];

        Arrays.fill(pos, -1);

        for (int i = 0; i < S.length(); i++) {
            pos[S.charAt(i) - 'a'] = i;
        }
        
        int N = T.length();
        Character[] C = new Character[N];
        
        for (int i = 0; i < N; i++) {
            C[i] = T.charAt(i);
        }
        
        Arrays.sort(C, new Comparator<Character>() {
            @Override
            public int compare(Character a, Character b) {
                return pos[a - 'a'] - pos[b - 'a'];
            }
        });
        
        StringBuilder sb = new StringBuilder();
        for (Character c : C) {
            sb.append(c);
        }
        return sb.toString();
    }
}

File: algorithms/leetcode/791-custom-sort-string/custom-sort-string.cpp
#include <algorithm>
#include <string>

class Solution {
public:
  std::string customSortString(std::string s, std::string t) {
    int len = s.size();

    int order[26] = {26};
    std::fill_n(order, 26, len);

    for (std::string::size_type i = 0; i < len; i++) {
      order[s[i] - 'a'] = i;
    }

    std::sort(t.begin(), t.end(), [&order](char a, char b) { return order[a - 'a'] < order[b - 'a']; });

    return t;
  }
};
